src/server/search/doc_index.h

```diff
@@ -52,10 +52,83 @@ struct SearchResult {
   std::optional<facade::ErrorReply> error;
 };
 
-template <typename T> using SearchField = std::pair<T /*identifier*/, T /*short name*/>;
+/* SearchField represents a field that can store combinations of identifiers and aliases in various
+   forms: [identifier and alias], [alias and new_alias], [new identifier and alias] (used for JSON
+   data) This class provides methods to retrieve the actual identifier and alias for a field,
+   handling different naming conventions and resolving names based on the schema. */
+class SearchField {
+ private:
+  static bool IsJsonPath(std::string_view name) {
+    if (name.size() < 2) {
+      return false;
+    }
+    return name.front() == '$' && (name[1] == '.' || name[1] == '[');
+  }
+
+ public:
+  SearchField() = default;
+
+  SearchField(StringOrView name, bool is_short_name)
+      : name_(std::move(name)), is_short_name_(is_short_name) {
+  }
+
+  SearchField(StringOrView name, bool is_short_name, StringOrView new_alias)
+      : name_(std::move(name)), is_short_name_(is_short_name), new_alias_(std::move(new_alias)) {
+  }
+
+  std::string_view GetIdentifier(const search::Schema& schema, bool is_json_field) const {
+    auto as_view = NameView();
+    if (!is_short_name_ || (is_json_field && IsJsonPath(as_view))) {
+      return as_view;
+    }
+    return schema.LookupAlias(as_view);
+  }
+
+  std::string_view GetShortName() const {
+    if (HasNewAlias()) {
+      return AliasView();
+    }
+    return NameView();
+  }
+
+  std::string_view GetShortName(const search::Schema& schema) const {
+    if (HasNewAlias()) {
+      return AliasView();
+    }
+    return is_short_name_ ? NameView() : schema.LookupIdentifier(NameView());
+  }
+
+  /* Returns a new SearchField instance with name and alias stored as views to the values in this
+   * SearchField */
+  SearchField View() const {
+    if (HasNewAlias()) {
+      return SearchField{StringOrView::FromView(NameView()), is_short_name_,
+                         StringOrView::FromView(AliasView())};
+    }
+    return SearchField{StringOrView::FromView(NameView()), is_short_name_};
+  }
+
+ private:
+  bool HasNewAlias() const {
+    return new_alias_.has_value();
+  }
+
+  std::string_view NameView() const {
+    return name_.view();
+  }
+
+  std::string_view AliasView() const {
+    return new_alias_.value().view();
+  }
+
+ private:
+  StringOrView name_;
+  bool is_short_name_;
+  std::optional<StringOrView> new_alias_;
```

Comment:
can new_alias be empty? I am asking because do we need this another level of indirection using optional?

Reply:
Yeah, I will fix that